File: go/fft_task_ref.go
//====================================================
// Go FFT Task Reference
//====================================================

package main

import (
	"fmt"
	"math"
	"math/cmplx"
	"math/rand"
	"time"
)

// generateSignal creates a sine wave with added noise
func generateSignal(n int, freq float64, noise []float64) []float64 {
	signal := make([]float64, n)
	dt := 1.0 / float64(n)
	
	for i := 0; i < n; i++ {
		t := float64(i) * dt
		sineWave := math.Sin(2 * math.Pi * freq * t)
		signal[i] = sineWave + noise[i]
	}
	return signal
}

// generateSyntheticData generates synthetic data with random noise
func generateSyntheticData(n int, freq, noiseAmp float64) []float64 {
	rand.Seed(time.Now().UnixNano())
	noise := make([]float64, n)
	for i := range noise {
		noise[i] = noiseAmp * (2*rand.Float64() - 1) // Random values between -noiseAmp and noiseAmp
	}
	return generateSignal(n, freq, noise)
}

// movingAverageFilter applies a moving average filter to the signal
func movingAverageFilter(windowSize int, signal []float64) []float64 {
	if windowSize <= 0 {
		return signal
	}
	
	n := len(signal)
	filtered := make([]float64, n-windowSize+1)
	
	// Calculate first window sum
	windowSum := 0.0
	for i := 0; i < windowSize; i++ {
		windowSum += signal[i]
	}
	filtered[0] = windowSum / float64(windowSize)
	
	// Sliding window for remaining points
	for i := 1; i < len(filtered); i++ {
		windowSum = windowSum - signal[i-1] + signal[i+windowSize-1]
		filtered[i] = windowSum / float64(windowSize)
	}
	
	return filtered
}

// basicDFT computes the Discrete Fourier Transform
func basicDFT(signal []float64) []complex128 {
	n := len(signal)
	spectrum := make([]complex128, n)
	
	for k := 0; k < n; k++ {
		sum := complex(0, 0)
		for j := 0; j < n; j++ {
			angle := -2 * math.Pi * float64(k*j) / float64(n)
			sum += complex(signal[j], 0) * cmplx.Rect(1, angle)
		}
		spectrum[k] = sum
	}
	return spectrum
}

// findPeaks identifies peaks above the threshold in the spectrum
func findPeaks(threshold float64, spectrum []complex128) []int {
	peaks := []int{}
	
	for i := 1; i < len(spectrum)-1; i++ {
		mag := cmplx.Abs(spectrum[i])
		if mag > threshold &&
			mag > cmplx.Abs(spectrum[i-1]) &&
			mag > cmplx.Abs(spectrum[i+1]) {
			peaks = append(peaks, i)
		}
	}
	return peaks
}

func main() {
	// Generate 10,000 samples with frequency 10Hz and noise amplitude 0.1
	samples := generateSyntheticData(10000, 10.0, 0.1)
	
	// Print first 10 samples
	fmt.Println("First 10 samples:")
	for i := 0; i < 10; i++ {
		fmt.Printf("%.6f ", samples[i])
	}
	fmt.Println()
	
	// Apply moving average filter
	filtered := movingAverageFilter(50, samples)
	
	// Compute FFT
	spectrum := basicDFT(filtered)
	
	// Find peaks above threshold (adjust threshold as needed)
	peaks := findPeaks(1000.0, spectrum)
	fmt.Printf("\nFound %d peaks in the spectrum\n", len(peaks))
}

File: julia/fft_task_ref.jl
"""
Julia FFT Task Reference
"""

"""
This program implements a signal processing pipeline:
1. Generate 10,000 samples of synthetic data: y(t) = sin(2πft) + noise
2. Apply a moving average filter (window size = 50)
3. Compute FFT
4. Find peaks above threshold
5. Output timing metrics and results in README.md
"""

using Random

"""
Generate synthetic signal with noise
"""
function generate_signal(n::Int, freq::Float64, noise::Vector{Float64})
    dt = 1.0 / n
    time_points = range(0, step=dt, length=n)
    sine_wave = sin.(2 * π * freq .* time_points)
    return sine_wave + noise[1:n]
end

"""
Generate synthetic data with random noise
"""
function generate_synthetic_data(n::Int, freq::Float64, noise_amp::Float64)
    noise = (2 * rand(n) .- 1) .* noise_amp
    return generate_signal(n, freq, noise)
end

"""
Apply moving average filter to signal
"""
function moving_average_filter(window_size::Int, signal::Vector{Float64})
    window_size <= 0 && return signal
    
    n = length(signal)
    filtered = zeros(n - window_size + 1)
    
    # Calculate first window sum
    window_sum = sum(signal[1:window_size])
    filtered[1] = window_sum / window_size
    
    # Sliding window for remaining points
    for i in 2:(n - window_size + 1)
        window_sum = window_sum - signal[i-1] + signal[i+window_size-1]
        filtered[i] = window_sum / window_size
    end
    
    return filtered
end

"""
Basic DFT implementation
"""
function basic_dft(signal::Vector{Float64})
    n = length(signal)
    spectrum = zeros(Complex{Float64}, n)
    
    for k in 0:(n-1)
        for j in 0:(n-1)
            omega = exp(-2π * im * k * j / n)
            spectrum[k+1] += signal[j+1] * omega
        end
    end
    
    return spectrum
end

"""
Find peaks above threshold in spectrum
"""
function find_peaks(threshold::Float64, spectrum::Vector{Complex{Float64}})
    n = length(spectrum)
    peaks = Int[]
    
    for i in 2:(n-1)
        mag = abs(spectrum[i])
        if mag > threshold &&
           mag > abs(spectrum[i-1]) &&
           mag > abs(spectrum[i+1])
            push!(peaks, i-1)  # 0-based indexing for consistency
        end
    end
    
    return peaks
end

function main()
    # Generate synthetic data
    samples = generate_synthetic_data(10000, 10.0, 0.1)
    println("First 10 samples: ", samples[1:10])
    
    # Example pipeline usage:
    # filtered_signal = moving_average_filter(50, samples)
    # spectrum = basic_dft(filtered_signal)
    # peaks = find_peaks(1.0, spectrum)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end 